encoding/ssh/filexfer/buffer.go

```diff
@@ -0,0 +1,268 @@
+func (b *Buffer) ConsumeUint64() (uint64, error) {
+	if b.Len() < 8 {
+		return 0, ErrShortPacket
+	}
+
+	v := binary.BigEndian.Uint64(b.b[b.off:])
+	b.off += 8
+	return v, nil
+}
+
+// AppendUint64 appends a single uint64 into the Buffer, in network byte order (big-endian).
+func (b *Buffer) AppendUint64(v uint64) {
+	b.b = append(b.b,
+		byte(v>>56),
+		byte(v>>48),
+		byte(v>>40),
+		byte(v>>32),
+		byte(v>>24),
+		byte(v>>16),
+		byte(v>>8),
+		byte(v>>0),
+	)
+}
+
+// ConsumeInt64 consumes a single int64 from the Buffer, in network byte order (big-endian) with two’s complement.
+// If Buffer does not have enough data, it will return ErrShortPacket.
+func (b *Buffer) ConsumeInt64() (int64, error) {
+	u, err := b.ConsumeUint64()
+	if err != nil {
+		return 0, err
+	}
+
+	return int64(u), err
+}
+
+// AppendInt64 appends a single uint64 into the Buffer, in network byte order (big-endian) with two’s complement.
+func (b *Buffer) AppendInt64(v uint64) {
+	b.AppendUint64(uint64(v))
+}
+
+// ConsumeByteSlice consumes a single string of raw binary data from the Buffer.
+// A string is a uint32 length, followed by that number of raw bytes.
+// If Buffer does not have enough data, or defines a length larger than available, it will return ErrShortPacket.
+func (b *Buffer) ConsumeByteSlice() ([]byte, error) {
+	length, err := b.ConsumeUint32()
+	if err != nil {
+		return nil, err
+	}
+
+	if b.Len() < int(length) {
+		return nil, ErrShortPacket
+	}
+
+	v := b.b[b.off:]
+	if len(v) > int(length) {
+		v = v[:length:length]
+	}
+	b.off += int(length)
+	return v, nil
+}
+
+// AppendByteSlice appends a single string of raw binary data into the Buffer.
+// A string is a uint32 length, followed by that number of raw bytes.
+func (b *Buffer) AppendByteSlice(v []byte) {
+	b.AppendUint32(uint32(len(v)))
+	b.b = append(b.b, v...)
+}
+
+// ConsumeString consumes a single string of UTF-8 encoded text from the Buffer.
+// A string is a uint32 length, followed by that number of raw bytes.
+// If Buffer does not have enough data, or defines a length larger than available, it will return ErrShortPacket.
+func (b *Buffer) ConsumeString() (string, error) {
+	v, err := b.ConsumeByteSlice()
+	if err != nil {
+		return "", err
+	}
+
+	return string(v), nil
+}
+
+// AppendString appends a single string of UTF-8 encoded text into the Buffer.
+// A string is a uint32 length, followed by that number of raw bytes.
+func (b *Buffer) AppendString(v string) {
+	b.AppendByteSlice([]byte(v))
+}
+
+// PutLength writes the given size into the first four bytes of the Buffer in network byte order (big endian).
+func (b *Buffer) PutLength(size int) {
+	if len(b.b) < 4 {
+		b.b = append(b.b, make([]byte, 4-len(b.b))...)
+	}
+
+	binary.BigEndian.PutUint32(b.b, uint32(size))
+}
+
+// MarshalBinary returns b as the binary encoding of b.
+func (b *Buffer) MarshalBinary() ([]byte, error) {
+	n := len(b.b)
+
+	// clamp cap() == len(), so any appends will reallocate.
```

Comment:
Appends will reallocate, but changes in the returned slice are still reflected in the Buffer.

Reply:
I would rather not allocate a buffer to return. It’s often sufficient. The `bytes.Buffer.Bytes()` function does the same. https://golang.org/src/bytes/buffer.go?s=2118:2149#L44

I did however, miss that this should be started from `b.b[b.off:]`. :laughing:

Reply:
If it starts at b.off, it's no longer MarshalBinary. It's Bytes.

Reply:
I mean, the two are functionally the same thing. They’re just named differently to match interfaces. Come to think of it though, “return b.Bytes()` is a good idea for the implementation as that was the intended version.